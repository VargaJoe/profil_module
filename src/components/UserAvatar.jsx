import React, { Component } 				from 'react';
import default_avatar 						from '../images/default_avater.svg';
import { connect }              			from 'react-redux';
import DATA                                 from '../config.json';

class UserAvatar extends Component {
	constructor(props) {
		super(props);
		this.state = {};
	}

	render () { 

		
		let picture = default_avatar;
		if ( this.props.userAvatar !== "" ) {
			
			picture = DATA.domain + this.props.userAvatar
		}

		

		return (
			<div className="user__avatar">
				<img src={picture} alt={this.props.userName}/>
			</div>
		)
	}
};

const mapStateToProps = (state) => {
	return {
		userAvatar :  	state.userImage.userImageSRC,
		userName : 		state.user.user.FullName
	};
};

export default connect(
	mapStateToProps
)(UserAvatar);
