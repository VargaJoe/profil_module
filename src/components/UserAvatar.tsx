import * as React							from 'react';
import { connect }              			from 'react-redux';
// import AvatarEditor 						from 'react-avatar-editor';
import Loader 								from './Loader';
// import Dropzone 							from 'react-dropzone';
import 'cropperjs/dist/cropper.css';

// save config 
const DATA = require('../config.json');

// save config 
const atob = require('atob');

export interface Props {
	onUpdate: Function;
	user: any;
}

export interface Point {
	x: number;
	y: number;
}

export interface IsChangedType {
	isChanged?: boolean;
	newImage?: any;
}

export interface State {
	rotate: number;
	image: any;
	allowZoomOut: boolean;
	position: Point;
	scale: number;
	borderRadius: number;
	preview: any;
	width: number;
	height: number;
	imageUpdates: IsChangedType;
}

export class UserAvatarComponent extends React.Component<Props, State> {

	// editor: AvatarEditor;

	constructor(props: Props) {
		super(props);
		this.state = {
			image: !this.props.user ? 
					DATA.domain + DATA.avatar + '/user.png' 
					: (
						!this.props.user.AvatarImageRef ? 
						DATA.domain + DATA.avatar + '/user.png' 
						:
						DATA.domain + this.props.user.AvatarImageRef.Path
					),
			allowZoomOut: false,
			position: { x: 0.5, y: 0.5 },
			scale: 1,
			rotate: 0,
			borderRadius: 300,
			preview: null,
			width: 250,
			height: 250,
			imageUpdates: {
				newImage: undefined,
				isChanged: false,
			}
		};
		this.useDefaultImage = 	this.useDefaultImage.bind(this);
		this.handleRotate = 	this.handleRotate.bind(this);
		// this.handleSave = 		this.handleSave.bind(this);
		this.b64toBlob = 		this.b64toBlob.bind(this);
		this.makeid = 			this.makeid.bind(this);
		this.imageChange = 		this.imageChange.bind(this);
	}

	// generate random names
	public makeid = () => {
		let text = '';
		let possible = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
		for (let index = 0; index < 7; index++) {
			text += possible.charAt(Math.floor( Math.random() * possible.length) );
		}
		return text;
	}

	// convert base64 to file
	public b64toBlob = (b64Data: string = '', sliceSize?: number) => {
		sliceSize = sliceSize || 512;
		if ( b64Data !== null) {
			let block = b64Data.split(';');
			let dataType = block[0].split(':')[1];
			let realData = block[1].split(',')[1];
			let filename = this.makeid() + '.' + dataType.split('/')[1];
			let byteCharacters = atob(realData);
			let byteArrays = [];
			for (let offset = 0; offset < byteCharacters.length; offset += sliceSize) {
				let slice = byteCharacters.slice(offset, offset + sliceSize);

				let byteNumbers = new Array(slice.length);
				for (let i = 0; i < slice.length; i++) {
					byteNumbers[i] = slice.charCodeAt(i);
				}
				let byteArray = new Uint8Array(byteNumbers);
				byteArrays.push(byteArray);
			}
			let blob = new Blob(byteArrays, {type: dataType});
			let newAvatar = new File([blob], filename);
			return newAvatar;
		} else {
			return '';
		}
	}

	// set image to gefault 
	public useDefaultImage() {
		this.setState({ image: !this.props.user ? 
			DATA.domain + DATA.avatar + '/user.png' 
			: (
				!this.props.user.AvatarImageRef ? 
				DATA.domain + DATA.avatar + '/user.png' 
				:
				DATA.domain + this.props.user.AvatarImageRef.Path
			) });
	}

	// rotate image to right 
	public handleRotate = (e: any) => {
		this.setState({
			rotate: parseFloat(e.target.value),
		});
	}

	// set cropper to the references
	// public setEditorRef = (editor: any) => {
	// 	if (editor) {
	// 	   this.editor = editor;
	//    	}
	// }

	// add image with dropp
	public handleDrop = (acceptedFiles: any) => {
		this.setState({ image: acceptedFiles[0] });
	}

	// get zoom value fron range
	public handleScale = (e: any) => {
		const scale = parseFloat(e.target.value);
		this.setState({ scale });
	}

	// detect all changes on image ( zoom, rotate )
	public imageChange = () => {
		this.setState({ 
			imageUpdates : {
				isChanged: true,
		}});
	}	

	// add nev image
	public handleNewImage = (e: any) => {
		let imgSrc;
		let reader = new FileReader();
		let that = this;
		reader.readAsDataURL(e.target.files[0]);
		reader.onload = function () {
			// get uploaded image 
			imgSrc = reader.result;
			that.setState({ image: imgSrc});
		};
	}

	// save changes
	// public handleSave = () => {
	// 	// is image is changed (add new, zoom, rotate)
	// 	if ( this.state.imageUpdates.isChanged) {
	// 		// convert to file
	// 		let imageFile = this.b64toBlob(this.editor.getImageScaledToCanvas().toDataURL());

	// 		this.setState({ imageUpdates : {
	// 			isChanged: true,
	// 			newImage: imageFile
	// 		}}, () => {
	// 			// add to EdutProfile component
	// 			this.props.onUpdate(this.state.imageUpdates);
	// 		});
	// 	}
	// }

	public render () {
		// if user is not updated then show loader
		if ( !this.props.user ) {
			return (<Loader/>);
		} else {
			return (
				<div className="user__avatar">
					<div className="user__avatar--blocks">
						<div>
							<h2>Avatar Editor</h2>
								<div>
									{/* <AvatarEditor
										ref={(ref: any) => this.setEditorRef(ref)}
										image={this.state.image}
										width={this.state.width}
										height={this.state.height}
										borderRadius={this.state.borderRadius}
										color={[0, 0, 0, 0.6]} // RGBA
										rotate={this.state.rotate}
										scale={this.state.scale}
										onImageChange={this.imageChange}
									/> */}
								</div>
							<div>
								<label htmlFor="newImage">New File:</label>
        						<input name="newImage" id="newImage" type="file" onChange={this.handleNewImage} />
							</div>
							<div>
								<label htmlFor="scale">Zoom:</label>
								<input 
									name="scale"
									id="scale"
									type="range"
									onChange={this.handleScale}
									min="0"
									max="2"
									step="0.01"
									defaultValue="1"
								/>
							</div>
							<div>
								<label htmlFor="rotate">Rotate:</label>
								<input 
									name="rotate"
									id="rotate"
									type="range"
									onChange={this.handleRotate}
									min="0"
									max="360"
									step="0.1"
									defaultValue="0"
								/>
							</div>
							<div>
								<br/>
								<button
									name="save"
									// onClick={this.handleSave}
									className="sn_btn"
								>
									Save Avatar
								</button>
							</div>
						</div>
						
					</div>	
				</div>
			);
		}
	}
}

const mapStateToProps = (state: any) => {
	return {
		user : state.user.user
	};
};

export default connect(
	mapStateToProps
)(UserAvatarComponent);
