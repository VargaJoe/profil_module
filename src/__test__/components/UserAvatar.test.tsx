jest.unmock('../../index.tsx');
jest.unmock('redux-mock-store');

import * as React 						  		from 'react';
import UserAvatar, { UserAvatarComponent } 		from '../../components/UserAvatar';
import { 
	    configure, 
		shallow,
		mount, 
		ShallowWrapper, 
		ReactWrapper } 							from 'enzyme';
import * as Adapter 					  		from 'enzyme-adapter-react-16';
import toJson 				      				from 'enzyme-to-json';
import * as createMockStore 					from 'redux-mock-store';
import { Provider } 							from 'react-redux';

configure( {adapter: new Adapter()} );
let base64 = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAQAAAC1+jfqAAAABGdBTUEAALGPC/xhBQAAACBjSFJNAAB6JgAAgIQAAPoAAACA6AAAdTAAAOpgAAA6mAAAF3CculE8AAAAAmJLR0QAAKqNIzIAAAAJcEhZcwAADdcAAA3XAUIom3gAAAAHdElNRQfiBR8NLAkf95jJAAAA5klEQVQoz33QP0uCcRTF8U9/1ElsKSgH91ZxewppMByMXILcfBdBb8ShuaVXIEW1SIM4SENTZfAMgkYNBYEQDZo8PyLvdO/3Hu65HMKKRBbWjdtF601DQ/kkWgkETSMPttz9d6EjEukm0aqi9VmfldfBhmNvMzZecmVP2xe4cI6GI5Cx75q0lq7CH7uCrpb073OxSrCuiDWToGigOp+qBorTdnmGeh5N5oKJJ73QMWckhQN1pIythUEdymg701C269K2b/dJi5oPfUMlJbG+T7Uw8FcvyvN5x7N3qaTgVC74KetkmsEP724z6YNUPMMAAAAldEVYdGRhdGU6Y3JlYXRlADIwMTgtMDUtMzFUMTM6NDQ6MDkrMDI6MDCmJO14AAAAJXRFWHRkYXRlOm1vZGlmeQAyMDE4LTA1LTMxVDEzOjQ0OjA5KzAyOjAw13lVxAAAABl0RVh0U29mdHdhcmUAd3d3Lmlua3NjYXBlLm9yZ5vuPBoAAAAASUVORK5CYII=';
describe('<UserAvatar /> rendering', () => {
		
	let store, 
		useravatarMount:  ReactWrapper<any, any>,
		userAvatarComponent: ShallowWrapper<any, any>;

	const mockStore = (createMockStore as any)();
	beforeEach( () => {
        store = mockStore({
			user: {
				user: {
					AvatarImageRef: 'something',
					Name: 'Test User'
				}
			}
		});
		const props = {
			user: {
				user: {
					AvatarImageRef: {
						Path: 'something'
					}
				}
			},
			onUpdate: Function
		};
        
		useravatarMount = mount(
            <Provider store={store}>
            	<UserAvatar {...props}/>
			</Provider>
		);
		let onUpdate = (value: any) => {
			return value;
		};
		let user = {
			AvatarImageRef: 'something',
			Name: 'Test User'
		};
		userAvatarComponent = shallow(
			<UserAvatarComponent onUpdate={onUpdate} user={user}/>
		);
	}); 
		
	// test Snapshot 
	it('Match to snapshot', () => {
		expect(toJson(useravatarMount)).toMatchSnapshot();
	});
		
	it('Test state', () => {
		expect(useravatarMount.children().children().instance().state.allowZoomOut).toBe(false);
		expect(useravatarMount.children().children().instance().state.scale).toBe(1);
		expect(useravatarMount.children().children().instance().state.rotate).toBe(0);
		expect(useravatarMount.children().children().instance().state.borderRadius).toBe(300);
		expect(useravatarMount.children().children().instance().state.preview).toBe(null);
		expect(useravatarMount.children().children().instance().state.width).toBe(250);
		expect(useravatarMount.children().children().instance().state.height).toBe(250);
	});

	it('Test props user data', () => {
		const props = {
			AvatarImageRef: 'something',
			Name: 'Test User'
		};
		expect(useravatarMount.children().children().props().user).toEqual(props);
	});

	it('Test props user data', () => {
		const props = {
			AvatarImageRef: 'something',
			Name: 'Test User'
		};
		expect(useravatarMount.children().children().props().user).toEqual(props);
	});

	it('test makeid', () => {
		const fff = useravatarMount.children().children();
		
		const spy = jest.spyOn(fff.instance() as any, 'makeid');
		let makedId = (fff.instance() as any).makeid();
		expect(spy).toHaveBeenCalledTimes(1);
		expect(typeof(makedId)).toBe('string');
	});

	it('test b64toBlob', () => {
		const fff = useravatarMount.children().children();
		const spy = jest.spyOn(fff.instance() as any, 'b64toBlob');
		let createFile = (fff.instance() as any).b64toBlob(base64);
		expect(spy).toHaveBeenCalledTimes(1);
		expect(spy).toBeCalledWith(base64);
		expect(typeof(createFile)).toBe('object');
		base64 = null;
		createFile = (fff.instance() as any).b64toBlob(base64);
		expect(spy).toBeCalledWith(base64);
		expect(createFile).toBe('');
	});

	it('test useDefaultImage', () => {
		const fff = useravatarMount.children().children();
		const spy = jest.spyOn(fff.instance() as any, 'useDefaultImage');
		(fff.instance() as any).useDefaultImage();
		expect(spy).toHaveBeenCalledTimes(1);
	});

	it('test handleRotate', () => {
		const fff = useravatarMount.children().children();
		const spy = jest.spyOn(fff.instance() as any, 'handleRotate');
		useravatarMount.setState({
			rotate: 0
		});
		const event = {target: {name: 'rotate', value: 19}};
		(fff.instance() as any).handleRotate(event);
		expect(spy).toHaveBeenCalledTimes(1);
		expect(spy).toBeCalledWith(event);
	});

	it('test handleDrop', () => {
		const fff = useravatarMount.children().children();
		const spy = jest.spyOn(fff.instance() as any, 'handleDrop');
		
		const files = {acceptedFiles: [ 'image']};
		(fff.instance() as any).handleDrop(files);
		// fff.find('#rotate').simulate('change', event);
		expect(spy).toHaveBeenCalledTimes(1);
		expect(spy).toBeCalledWith(files);
	});

	it('test handleScale', () => {
		const fff = useravatarMount.children().children();
		const spy = jest.spyOn(fff.instance() as any, 'handleScale');
		
		const event = {target: {name: 'scale', value: 19}};
		(fff.instance() as any).handleScale(event);
		// fff.find('#rotate').simulate('change', event);
		expect(spy).toHaveBeenCalledTimes(1);
		expect(spy).toBeCalledWith(event);
	});

	it('test imageChange', () => {
		const fff = useravatarMount.children().children();
		const spy = jest.spyOn(fff.instance() as any, 'imageChange');

		(fff.instance() as any).imageChange();
		// fff.find('#rotate').simulate('change', event);
		expect(spy).toHaveBeenCalledTimes(1);
	});

	it('test handleSave', () => {
		const fff = useravatarMount.children().children();
		const spy = jest.spyOn(fff.instance() as any, 'handleSave');
		(fff.instance() as any).handleSave();
		// fff.find('#rotate').simulate('change', event);
		expect(spy).toHaveBeenCalledTimes(1);
	});

	it('test setEditorRef', () => {
		const fff = useravatarMount.children().children();
		const spy = jest.spyOn(fff.instance() as any, 'setEditorRef');
		let editor: any = 'something';
		(fff.instance() as any).setEditorRef(editor);
		// fff.find('#rotate').simulate('change', event);
		expect(spy).toHaveBeenCalledTimes(1);
	});
		
}); 