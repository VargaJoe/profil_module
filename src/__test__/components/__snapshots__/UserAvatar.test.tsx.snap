// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserAvatar /> rendering Match to snapshot 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(UserAvatarComponent)
    onUpdate={[Function]}
    user={
      Object {
        "user": Object {
          "AvatarImageRef": Object {
            "Path": "something",
          },
        },
      }
    }
  >
    <UserAvatarComponent
      dispatch={[Function]}
      onUpdate={[Function]}
      user={
        Object {
          "AvatarImageRef": "something",
          "Name": "Test User",
        }
      }
    >
      <div
        className="user__avatar"
      >
        <div
          className="user__avatar--blocks"
        >
          <div>
            <h2>
              Avatar Editor
            </h2>
            <div />
            <div>
              <label
                htmlFor="newImage"
              >
                New File:
              </label>
              <input
                id="newImage"
                name="newImage"
                onChange={[Function]}
                type="file"
              />
            </div>
            <div>
              <label
                htmlFor="scale"
              >
                Zoom:
              </label>
              <input
                defaultValue="1"
                id="scale"
                max="2"
                min="0"
                name="scale"
                onChange={[Function]}
                step="0.01"
                type="range"
              />
            </div>
            <div>
              <label
                htmlFor="rotate"
              >
                Rotate:
              </label>
              <input
                defaultValue="0"
                id="rotate"
                max="360"
                min="0"
                name="rotate"
                onChange={[Function]}
                step="0.1"
                type="range"
              />
            </div>
            <div>
              <br />
              <button
                className="sn_btn"
                name="save"
                onClick={[Function]}
              >
                Save Avatar
              </button>
            </div>
          </div>
        </div>
      </div>
    </UserAvatarComponent>
  </Connect(UserAvatarComponent)>
</Provider>
`;
